android/libraries/rib-android/src/main/kotlin/com/uber/rib/core/RibActivity.kt

```diff
@@ -241,3 +243,10 @@ abstract class RibActivity :
       }
   }
 }
+
+private val <T : Comparable<T>> LifecycleScopeProvider<T>.lifecycleRange: ClosedRange<T>
+  get() {
```

Comment:
Is the idea to just catch the error earlier than when it's launcehd in unstarted ones? Do we need to handle other complexity of the original Autodispose resolveScopes method?

Reply:
Yes! We must throw at call site to keep the behavior. I don't thought thoroughly whether we *should*, mostly worried about retro-compatibility, but this one may have good reasons to throw at call site instead of in `Completable`.

We also throw early if pass in a range of active ranging from, e.g. (`STARTED`..`STOPPED`) but `SharedFlow` is still in `CREATED` (note on current implementation it should never happen because we start the range with `peekLifecycle()`):
```kotlin
internal fun <T : Comparable<T>> SharedFlow<T>.asScopeCompletable(
  range: ClosedRange<T>,
  context: CoroutineContext = EmptyCoroutineContext,
): CompletableSource {
  ensureAlive(range)
  return rxCompletable(RibDispatchers.Unconfined + context) { first { it == range.endInclusive } }
}

private fun <T : Comparable<T>> SharedFlow<T>.ensureAlive(range: ClosedRange<T>) {
  val lastEmitted = replayCache.lastOrNull()
  when {
    lastEmitted == null || lastEmitted < range.start -> throw LifecycleNotStartedException()
    lastEmitted >= range.endInclusive -> throw LifecycleEndedException()
  }
}
```

Reply:
> Do we need to handle other complexity of the original Autodispose resolveScopes method?

I don't think so. I think we got this covered. As I mentioned, it's a hopefully less complex impl that achieves the same behavior (and does not depend on Flow->Observable interop).